presto-main/src/main/java/com/facebook/presto/server/StatementResource.java

```diff
@@ -185,20 +210,25 @@ public Response createQuery(
     @GET
     @Path("{queryId}/{token}")
     @Produces(MediaType.APPLICATION_JSON)
-    public Response getQueryResults(
+    public void getQueryResults(
             @PathParam("queryId") QueryId queryId,
             @PathParam("token") long token,
             @QueryParam("maxWait") Duration maxWait,
-            @Context UriInfo uriInfo)
+            @Context UriInfo uriInfo,
+            @Suspended AsyncResponse asyncResponse)
             throws InterruptedException
     {
         Query query = queries.get(queryId);
         if (query == null) {
-            return Response.status(Status.NOT_FOUND).build();
+            asyncResponse.resume(Response.status(Status.NOT_FOUND).build());
+            return;
         }
 
-        Duration wait = WAIT_ORDERING.min(MAX_WAIT_TIME, maxWait);
-        return createResponse(query, query.getResults(token, uriInfo, wait));
+        Duration wait = WAIT_ORDERING.min(MAX_WAIT_TIME, firstNonNull(maxWait, DEFAULT_WAIT_TIME));
+        CompletableFuture<Response> future = query.getResults(token, uriInfo, wait);
+
+        bindAsyncResponse(asyncResponse, future, responseExecutor)
+                .withTimeout(new Duration(5, SECONDS));
```

Comment:
Good catch, I just put this in here while testing. We should do similar to `TaskResource` where we take the expected wait time plus 5 seconds as a hard timeout.